SyncKotlin/app/src/main/java/com/google/android/gms/fit/samples/synckotlin/FitSync.kt

```diff
@@ -0,0 +1,227 @@
+
+        val fitnessOptions: FitnessOptions = FitnessOptions.builder()
+            .addDataType(DataType.TYPE_STEP_COUNT_DELTA, FitnessOptions.ACCESS_WRITE)
+            .build()
+    }
+
+    private val notifications = Notifications(appContext)
+
+    private val dataSource = DataSource.Builder()
+        .setDataType(DataType.TYPE_STEP_COUNT_DELTA)
+        .setAppPackageName(appContext)
+        .setType(DataSource.TYPE_RAW)
+        .build()
+
+    private val sharedPreferences = appContext.getSharedPreferences(
+        appContext.getString(R.string.preference_file_key),
+        Context.MODE_PRIVATE
+    )
+
+    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
+
+    override suspend fun doWork(): Result {
+        if (!isActivityRecognitionPermissionApproved(appContext)) {
+            return buildRetryOrNotifyFailureResult(
+                appContext.getString(R.string.error_android_permissions))
+        } else if (!isOAuthPermissionsApproved(appContext, fitnessOptions)) {
+            return buildRetryOrNotifyFailureResult(
+                appContext.getString(R.string.error_oauth_permissions))
+        }
+
+        try {
+            performSync(getGoogleAccount(appContext, fitnessOptions))
+        } catch (e: Exception) {
+            return buildRetryOrNotifyFailureResult(e.toString())
+        }
+
+        sharedPreferences.edit(commit = true) {
+            putString(
+                appContext.getString(R.string.last_sync_date),
+                getCurrentDateTimeString()
+            )
+        }
+
+        /*
+          Repeated work is achieved by using a one-time work request, where the final action is to
+          enqueue another work request, as seen here, setting the appropriate initial delay.
+        */
+        enqueueRepeatedJob(appContext)
+        return Result.success()
+    }
+
+    private fun buildRetryOrNotifyFailureResult(message: String): Result {
+        return when (runAttemptCount) {
+            MAX_RUN_ATTEMPTS -> {
+                sharedPreferences.edit {
+                    putBoolean(
+                        appContext.getString(R.string.last_sync_failed),
+                        true
+                    )
+                }
+                notifications.showNotification(message)
+                Result.failure(errorData(message))
+            }
+            else -> Result.retry()
+        }
+    }
+
+    private fun performSync(account: GoogleSignInAccount) {
+        Log.i(TAG, "Performing sync")
+        // Get the last sync time
+        val lastSyncMillis =
+            sharedPreferences.getLong(appContext.getString(R.string.last_sync_timestamp), 0)
+
+        // Get any steps since lastSyncMillis. StepsDatabase is a class just for the purposes of
+        // this sample, to serve as a placeholder for the storage an app might use, be it local or
+        // in a app-specific cloud.
+        val db = StepsDatabase()
```

Comment:
Optional: I usually prefer to always use full names, so it's clear to all devs, e.g., 'database'... some might not know what that stands for.